engine/buildkit/client.go

```diff
@@ -452,6 +492,121 @@ func withDescHandlerCacheOpts(ctx context.Context, ref bkcache.ImmutableRef) con
 	})
 }
 
+func (c *Client) ListenHostToContainer(
+	ctx context.Context,
+	hostListenAddr, proto, upstream string,
+) (*session.ListenResponse, func() error, error) {
```

Comment:
I think we'll want to preprend this method with the boilerplate of the other methods:
```go
ctx, cancel, err := c.withClientCloseCancel(ctx)
 	if err != nil {
 		return "", "", nil, err
 	}
 	defer cancel()
```

Which helps ensure that it stops if the server is shutdown. I can't tell 100% if it's strictly necessary for this exact method, but I *think* worst case it's harmless.

I needed to add that because there were cases where the original client exited but other nested clients were still trying to do stuff and ended up leaking and/or blocking shutdown.